### DISCUSSION TITLE
Performance de l'instanciation du moteur

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#256] [General]
Je rencontre des problèmes de performance liés à publicodes sur mesaidesvélo. Cela concerne : 

- le temps de calcul d'une page d'aide (je n'ai pas encore creusé le sujet)
- le temps initial d'instanciation du moteur avec les règles `new Publicodes(aidesVelo)` (le sujet de ce message)

Le temps d'instanciation est actuellement de **~500ms**, ce qui retarde d'autant le “First Input Delay” qui est un core web vital de Google. Autrement dit, le temps que tout le Javascript s'initialise, l'application ne répond pas. Le [fichier publicodes source](https://github.com/mquandalle/mesaidesvelo/blob/eb36ddb729ad1a78fdeaa7f90e207adbf4187aac/src/aides.yaml) fait environ 5000 lignes.

Le déplacement de cette logique d'instanciation dans un worker n'est pas facile car elle rend tous les appels asynchrones, ce qui complique la gestion de l'UI (ne fonctionne pas en SSR, pas de mode concurrent en Svelte comme en React).

Mais fondamentalement le temps d'instanciation initial ne devrait pas être si long, je cherche donc à le réduire.

Une piste qu'on évoque depuis longtemps est de déplacer "coté serveur" l'appel à `parsePublicodes` lors du build de l'application. En effet les analyses et transformations sont toujours les mêmes pour une base de règles donnée (résolution des noms, inférence d'unité, substitution des remplacements, etc.) et n'ont pas besoin d'être refaits par chaque navigateur.

Malheureusement ces transformations augmentent aussi beaucoup la taille de la structure de donnée utilisée :

```js
const jsonSource = yaml.parse(aidesVeloTextSource); // 100ms (lors du build)
const parsedPublicodes = parsePublicodes(jsonSource): // 500ms (lors du premier chargement)
```

|Taille | Sans Gzip | Avec Gzip|
|--- | --- | ---|
|`jsonSource` | 121 Ko | 21 Ko|
|`parsedPublicodes` | 18 Mo | 662 Ko|
|ratio | × 152 | × 31|

Une fois sérialisé avec une `JSON.stringify`, `parsedRules` fait 18Mo soit 150 fois plus que le `jsonSource`. Si on `gzip` le ratio s'améliore car le fichier contient beaucoup de redites, mais on reste quand même 30 fois plus lourd que le `jsonSource`. En l’état cette structure de données `parsedPublicodes` est beaucoup trop lourde pour être envoyé du serveur au client. (à noter que la taille sans Gzip a aussi son importance car ça correspond aux octets que le navigateur va devoir parser)

@johangirod me faisait remarquer que la structure contient des objets Javascript référencés à plusieurs endroits, et qui sont alors dupliqués lors d’un `JSON.strinfigy` naïf. On peut les envoyer qu'une seule fois et utiliser des références ailleurs. J'ai testé cette méthode via la librairie [`devalue`](https://github.com/rich-harris/devalue) qui compile un objet Javascript en un code source de fonction javascript qui va le reconstruire, exemple `(function() { a = {x: 1}, b = {y: a}; ... })()`. Cette sérialisation donne un meilleur résultat mais qui reste trop lourd pour être envoyé du serveur au client :

|Taille | Sans Gzip | Avec Gzip|
|--- | --- | ---|
|`jsonSource` | 121 Ko | 21 Ko|
|`parsedPublicodes` | 3,1 Mo | 445 Ko|
|ratio | × 25 | × 21|

Avec cette méthode le fichier envoyé au client reste 20 fois plus lourd que d'envoyer le `jsonSource`.

J'envisage de recoder depuis zéro un parseur et un interpréteur limité à mon cas d'usage (sans certaines fonctionnalités de publicodes comme les recalculs, substitution de références, désactivation de parent) mais en les gardant en tête sur les structures de données utilisées. L’idée est d'envoyer au client mon fichier d'aides compilé dans un format
- pas lourd (~120% du fichier source gzippé)
- rapide à charger (~50ms)

J'y réfléchis encore un peu avant de me lancer.

**Comment by mquandalle** [comment]
Le problème se pose aussi sur mon-entreprise mais je n'ai pas creusé
https://pagespeed.web.dev/report?url=https%3A%2F%2Fmon-entreprise.urssaf.fr%2Fsimulateurs%2Fsalaire-brut-net&form_factor=mobile

**Reply by laem** [reply]
On a une PR qui traite plus ou moins de ça pile en ce moment sur nosgestesclimat.fr : https://github.com/datagir/nosgestesclimat-site/pull/629. 

En gros, on attendait le chargement des règles avant d'affichier le site. Grosse erreur. Maintenant qu'on l'a résolue, il y a un petit lag d'instanciation du moteur, qui est ton sujet ici. 

De notre côté (pas autant de règles ?) trop peu perceptible pour être un problème, mais idéalement faudrait aussi le résoudre.

**Reply by laem** [reply]
Comment expliquer le bon score FIP de nosgestesclimat.fr ? 

![image](https://user-images.githubusercontent.com/1177762/188841860-4e9de4a9-c61f-41cb-b720-c5b85c908cf8.png)

**Reply by mquandalle** [reply]
Hum je ne sais pas, sur https://pagespeed.web.dev/report?url=https%3A%2F%2Fnosgestesclimat.fr%2Fsimulateur%2Fbilan, il y a à la fois les données terrain en haut (où votre FID est ok) et les données labo en dessous (ou les vitaux sont au rouge).

Hypothèse : vu que les données de terrain sont mesurées en "conditions réelle", peut-être que tous les utilisateurs arrivent sur cette page via navigation in-app avec l'application déjà chargée, et il ne s'agit pas d'premier chargement depuis l'URL comme dans l'exemple "laboratoire".

**Reply by laem** [reply]
aaaah oui, c'est clairement ça, bien vu.

**Comment by laem** [comment]
> J'envisage de recoder depuis zéro un parseur et un interpréteur limité à mon cas d'usage (sans certaines fonctionnalités de publicodes comme les recalculs, substitution de références, désactivation de parent) mais en les gardant en tête sur les structures de données utilisées. L’idée est d'envoyer au client mon fichier d'aides compilé dans un format

As-tu pu identifier quelles étapes prennent du temps ? Quelles étapes consomment beaucoup d'espace ? Car c'est peut-être juste la génération et la gestion mémoire d'un 18 Mo qui pose ces défis de temps de traitement :o Potentiellement, tu auras enlevé des fonctionnalités, mais le problème structurel ne serait pas touché.

**Reply by mquandalle** [reply]
J'ai mesuré chacune des sous-étapes de [parsePublicodes](https://github.com/betagouv/publicodes/blob/master/packages/core/source/parsePublicodes.ts#L64) et globalement il n'y avait pas de gros déséquilibre, chaque étape prenait autour de 100ms.

(18Mo c'est la taille sérialisée, avec les mêmes objets JS qui apparaissent à plusieurs endroits de la structure, la taille en mémoire c'est plutôt 3Mo du coup (à vérifier avec les outils de débogage). Mais clairement c'est beaucoup, je pense parce qu'on a pas suivi l'évolution de cette structure de données au fur et à mesure des modifications.)

Je ne me suis pas encore lancé mais l'autre aspect qui m'intéressait dans le fait de repartir d'une version minimale du moteur, c'est de pouvoir avoir des tests benchmarks #21, et de vérifier qu'on ne “paye que ce qu'on utilise”. Exemple si je n'utilise pas la désactivation de branche ou les remplacements ou les recalcul, la taille des règles parsées envoyées au client et le temps de chargement ne doivent pas être pénalisés par la ré-introduction de ces fonctionnalités.

**Comment by laem** [comment]
Pour ton sujet précis de mes aides vélo : pourquoi tu n'instancies pas juste le périmètre de la ville consultée ? https://mesaidesvelo.fr/ville/nantes 
Juste Nantes + ce qu'il y a au-dessus mais pas les autres villes.

**Reply by mquandalle** [reply]
Oui je vais aller vers ça. Mais
- il y a quand même un gros problème de perf à régler de base
- il faut que publicodes supporte “l'ajout de règles à chaud”
- dans l'implémentation actuelle, [certaines aides sont référencés dans une liste globale](https://github.com/mquandalle/mesaidesvelo/blob/afbcafab576de281c1e5f98ef09bf0b2ea0c12a5/src/aides.yaml#L4351), si on les enlève ça casse

**Reply by laem** [reply]
> il faut que publicodes supporte “l'ajout de règles à chaud"

Est-ce que tu as un avis sur la difficulté de cette fonctionnalité ?

**Comment by laem** [comment]
> Le déplacement de cette logique d'instanciation dans un worker n'est pas facile car elle rend tous les appels asynchrones, ce qui complique la gestion de l'UI (ne fonctionne pas en SSR, pas de mode concurrent en Svelte comme en React).

Et ça ne te semble pas faisable de parser dans un worker, puis d'envoyer le résultat au thread principal, qui attendrait sans freeze puis pourrait ensuite tout évaluer en synchrone ?

**Reply by mquandalle** [reply]
Ça doit être faisable, mais pas facile :

- pas de primitive “suspense” en Svelte, on ne peut pas facilement suspendre le rendu pour attendre un résultat asynchrone (c'est possible mais pas ergonomique si on doit le faire à chaque `evaluate`)
- l'approche “faux-synchrone” que tu décris pourrait fonctionner, mais il faut quand même un peu de logique : collecter les objectifs à calculer dans un état global, et exposer ensuite les résultats sous la forme d'un contexte
- pour le SSR il faut garder une approche 100% synchrone, et le SSR est important pour que Google référence bien des pages comme https://mesaidesvelo.fr/ville/nantes. Donc passer en asynchrone ça serait seulement côté client, pas côté serveur, ça complique encore beaucoup le code

Pour moi c'est difficile d'envisager de faire de l’asynchrone sans abstraction type "suspense" conçue pour gérer ça correctement dans le rendu serveur ou client.

Et puis surtout le calcul à la fin c'est `Math.min(40% * vélo . prix, 500)`, ça devrait pas faire exploser nos ordis 😄

**Reply by laem** [reply]
Je vais peut-être essayer une approche worker pour NGC, je te dirai.

**Reply by laem** [reply]
Bon clairement passer l'engine instancié n'est pas possible dans un WebWorker. 
J'arrive pas encore à savoir si Comlink permet de faire ça. Edit : si, tu l'as fait, mais ça rend tout asynchrone : https://github.com/betagouv/mon-entreprise/pull/1349/files
À côté de ça, cette technique est très intéressante : https://codesandbox.io/s/usetransitioneffect-canvas-example-m0hhf9?from-embed=&file=/src/App.tsx

Mais ça supposerait de pouvoir réécrire le parse pour qu'il puisse être interrompu comme un générateur. T'avais investigué ça @mquandalle  ?

**Reply by laem** [reply]
J'avais l'impression que le problème venait surtout du parsing, plus que de l'evaluate ?

**Reply by mquandalle** [reply]
Oupsi j'ai supprimé mon message, mais pour te répondre non je n'ai pas investigué l'idée d'un parseur interruptible.

**Reply by johangirod** [reply]
Pour info, on a testé cette approche sur mon-entreprise, et ç'était beaucoup trop compliqué à mettre en oeuvre dans un projet existant. En from scratch, pourquoi pas, mais sur un projet avec des appels à evaluate non async, trop compliqué.

CF https://github.com/betagouv/mon-entreprise/pull/2744

D'accord avec le constat : il faut adresser le problème à la racine plutôt que de le déplacer dans un worker.

**Comment by laem** [comment]
Ici on a un cas relativement lié à cette issue : on a un modèle "parc français" qui n'a aucun intérêt à être recalculé côté client. Il ne contient pas de questions. 

https://github.com/datagir/nosgestesclimat/pull/1433/files

On pourrait donc inscrire une valeur "valeur par défaut précalculée". Aucune raison alors de parser ce modèle sauf à le demander explicitement. 

C'est qu'une idée en l'air, il faut que je la développe.

**Reply by laem** [reply]
Au-delà de ces valeurs par défaut, on parse la moitié de notre modèle (nosgestesclimat.fr/actions) bien avant d'arriver sur cette page, c'est bête.

**Reply by johangirod** [reply]
Autre sujet lié de notre côté : on historise au fur et à mesure la base de règle. Cela veut dire que les années passant, on va se retrouver avec plus en plus de code dédié à des calculs qui ne seront pas effectués par défaut côté client. Il serait intéressant de pouvoir passer une base de règle qui concerne uniquement une date donnée, en précalculant tout ce qui peut l'être vis à vis de cette date.

On peut imaginer que ce soit le cas avec d'autres variables également. En fait il y a un peu l'idée d'exécuter au build tout ce qui est statique pour envoyer un publicodes plus compact et alléger l'évaluation / le parsing côté client.

**Reply by laem** [reply]
Ah oui très intéressant comme cas d'usage, propre à la legislation.

**Comment by EmileRolley** [comment]
Lié à ce sujet : https://publicodes.github.io/publicodes-tools/modules/optims.html

**Comment by johangirod** [comment]
Une piste a explorer serait l'utilisation d'un AST sous forme de tableau :
https://www.cs.cornell.edu/~asampson/blog/flattening.html

Utilisé en JS ici avec succès :
https://calendar.perfplanet.com/2023/benchmarking-profiling-and-optimizing-javascript-libraries/
